---
output:
  github_document:
    df_print: kable
  pdf_document: default
  word_document: default
  html_document: default
---

# Mock community analysis

```{r setup, warning=F, message=F}
library(phyloseq)
library(dplyr)
library(tibble)
library(ggplot2)
theme_set(theme_bw())

source(file.path("..", "..", "R", "read_input.R"))
```

## Read data

First, we read all the data by simply specifying the results directory path:

```{r}
res_base <- file.path("..", "results")
# UNOIS3 pipeline
res_unoise <- read_pipeline_results(file.path(res_base, "unoise", "data"))
```

The result of `read_pipeline_results` is a list with different components: *`r paste(names(res_unoise), collapse=', ')`*.

- `otutab`: Count matrix with ASVs/OTUs in rows and samples in columns.
- `taxonomy`: List of data frames, one per taxonomic assignment method (here, only one was used: `r paste(names(res_unoise$taxonomy), collapse=', ')`) with OTU, taxonomic ranks and a `def_rank` column specifying the rank, at which the names are defined. The taxonomic table should never have any `NA` values. Unknown names were auto-filled based on the known names, e.g. `Ascomycota_ord_unknown`. At the species level, taxa re named `Saccharomyces sp.` (if genus known) or `Ascomycota clone` (genus unknown). These name patterns can be configured in `read_pipeline_results` (`unknown_fmt` and `unknown_species_fmt`, see documentation of `replace_missing_taxa`). Example with shortened names:
```{r echo=F}
tax_example <- res_unoise$taxonomy[[1]]
for (i in 1:6) {
  tax_example[, i] <- substr(tax_example[, i], 1, 6)
}
tax_example[, 7] <- substr(tax_example[, 7], 1, 15)
head(tax_example)
```
- `refseq`: OTU/ASV sequences as `Biostrings::DNAStringSet` object.
- `tree`: tree object (not present in this case since the ITS region is too variable for creating a phylogenetic tree)
- `cmp`: List of sequence comparisons with OTUS/ASVs. In this case, the ITS2 sequences of the mixed species in the mock communities were compared. The first few rows: 
```{r echo=F}
head(res_unoise$cmp$mock)
```
- `itsx_results`: Finally, the rDNA domains were searched using ITSx, since these are fungal ITS sequences. The `ITSx_cat` contains (shortened) messages from ITSx. *partial_5.8S* means that not the whole 5.8S region was found, which is not surprising since the forward primer is located on the 5.8S region. The first few rows: 
```{r echo=F}
head(res_unoise$itsx_results)
```

*Note*: Every taxonomy data frame has a 'ranks' attribute. This can be useful and is also needed later by `make_phyloseq`.

```{r}
attr(res_unoise$taxonomy$`unite_all-sintax_70`, 'ranks')
```


## Mixed vs. amplified frequencies

As an example, we compare the relative mixed amounts (genomic DNA) of the different species with their read counts in the `mock1` community:

```{r}
# we analyse the results of the UNOISE3 pipeline
res <- res_unoise

counts <- res$otutab %>%
  as.data.frame() %>%
  rownames_to_column("OTU")

mock_data <- read.delim(file.path("..", "mock", "mock_data.txt")) %>%
  # use the sequence comparison the match obtain the ASV names for each species
  left_join(res$cmp$mock, c("name" = "target")) %>%
  # then add the ASV counts
  left_join(counts, c("query" = "OTU")) %>%
  # then sum up the ASV counts per species
  group_by(name, rel_abund_m1) %>%
  summarise(
    n_reads = sum(mock1),
    n_reads = ifelse(is.na(n_reads), 0, n_reads),
    ASVs = paste(unique(na.omit(query)), collapse = ","),
    n_asvs = n(),
    .groups = "drop"
  ) %>%
  arrange(-rel_abund_m1)

head(mock_data)
```

```{r fig.width=6, fig.height=3, message=F, warning=F}
ggplot(mock_data, aes(rel_abund_m1, n_reads)) +
  geom_point(aes(size = as.factor(n_asvs)), pch = 21, alpha = 0.7) +
  geom_smooth(method = "lm", formula = "y~x", se = F, size = 0.3) +
  scale_size_discrete(range = c(1, 4)) +
  scale_x_log10() +
  scale_y_log10() +
  labs(
    x = "Mixed relative concentration",
    y = "Read count",
    size = "Number of ASVs"
  ) +
  theme_bw() +
  theme(strip.text.y = element_text(angle = 0))
```

## Make phyloseq object

The `make_phyloseq` function in the *read_input.R* script can directly create a phyloseq object. The only additional thing we *must* supply here is a metadata table (data frame) with 'sample' column or the row names set to samples.

```{r}
# minimal metadata table
meta <- data.frame(sample = c("mock1", "mock2"))
# create the phyloseq object
physeq <- make_phyloseq(res_unoise, meta = meta)
physeq
```

Since there can be several taxonomic assignments, `make_phyloseq` just chooses the first one. To choose a specific one, supply the taxonomy table using the `taxonomy` argument. In this example, we modify the taxonomy by removing the `def_rank` column first.

```{r}
tax <- res_unoise$taxonomy$`unite_all-sintax_70` %>%
  select(-def_rank)
physeq <- make_phyloseq(res_unoise, meta = meta, taxonomy = tax)
```

```{r fig.width=7.5, fig.height=3.2, message=F, warning=F}
plot_bar(physeq, fill = "species") +
  guides(fill = guide_legend(ncol = 2))
```

